Fixes

Hroy02 opened the pull request.

Hroy02 committed:
Fix Squad and Future

Hroy02 committed:
Merge branch 'master' into Fixes

Hroy02 committed:
Fix TimeService to use TerminationTickBroadcast

Hroy02 committed:
Create Queue at new event/broadcast subscription

Hroy02 committed:
Merge branch 'master' into Fixes

Hroy02 committed:
Refix Squad and Future

IdanTo1 commented on src/main/java/bgu/spl/mics/Future.java:
```diff
@@ -76,13 +76,12 @@ public boolean isDone() {
-            } catch (InterruptedException e) {
-                return null;
-            }
+    public synchronized T get(long timeout, TimeUnit unit) {
```
why is it needed? two threads will never acsses the same Future

IdanTo1 replied:
got it. for wait you need synchronized, my bad

IdanTo1 commented on src/main/java/bgu/spl/mics/MessageBrokerImpl.java:
```diff
@@ -38,6 +38,7 @@ private MessageBrokerImpl() {
 	 @post: m's queue will receive Event<T> objects.
 	 */
 	public <T> void subscribeEvent(Class<? extends Event<T>> type, Subscriber m) {
+		if(_eventSubscribers.get(type) == null) _eventSubscribers.put(type, new ConcurrentLinkedQueue<>());
```
will never happen
one should not subscribe to event before it was registered. the bug is not here but in an other place

IdanTo1 replied:
my bad again. you create a queue for a broadcast and not a subscriber. good job.
but, add a comment about it being thread-safe

IdanTo1 commented on src/main/java/bgu/spl/mics/MessageBrokerImpl.java:
```diff
@@ -47,6 +48,7 @@ private MessageBrokerImpl() {
 	 * @post: m's queue will receive Broadcast<T> objects.
 	 */
 	public void subscribeBroadcast(Class<? extends Broadcast> type, Subscriber m) {
+		if(_broadcastSubscribers.get(type) == null) _broadcastSubscribers.put(type, new ConcurrentLinkedQueue<>());
```
same

IdanTo1 commented on src/main/java/bgu/spl/mics/application/passiveObjects/Squad.java:
```diff
@@ -102,8 +102,8 @@ public boolean getAgents(List<String> serials) throws InterruptedException {
                     agent.wait();
-                    agent.acquire();
                 }
+                agent.acquire();
```
cool, my bad!

IdanTo1 commented on src/main/java/bgu/spl/mics/application/publishers/TimeService.java:
```diff
@@ -48,7 +48,7 @@ public void run() {
             _currentTick++;
         }
-        getSimplePublisher().sendBroadcast(new TerminationTick());
+        getSimplePublisher().sendBroadcast(new TerminationTickBroadcast());
```
my bad either :sweat_smile:

IdanTo1 committed:
fix a bug - MessageBrokerTest didn't call subscribers' callback

Signed-off-by: Idan Tomer <idan.tomer2@gmail.com>

Hroy02 committed:
Fix MessageBrokerImpl

1. Use take instead of poll because poll isn't properly blocking
2. Remove all references to unregistering subscriber

Hroy02 committed:
Fix MessageBrokerTest

1. Remove redundant registering
2. Fix testInterruption to use proper instance of new subscriber

IdanTo1 committed:
Merge remote-tracking branch 'origin/master' into Fixes
